### DISCUSSION TITLE
Does Componenet can have multiple buttons?

### DISCUSSION THREAD

**Discussion by doowonee** [flame-engine/flame#1883] [Q&A] [answered]
Hello I'm making `class HeaderContainer extends PositionComponent` which has 3 `SpriteButtonComponent` as fields. But When I click the button inside MyComponenet `onPressed` does not work. I add a log inside the anonymous function and it showed me nothing.

When I add three buttons directly in `FlameGame` and It works. I don't understand Why child component `HeaderContainer` can not propagate click event to child components

Here is `FlamgeGame.onLoad()`

```
   add(HeaderContainer()
      ..positionType = PositionType.viewport
      ..size = Vector2(size.x, 60));
```

And `HeaderContainer.onLoad()`

```
class HeaderContainer extends PositionComponent
    with HasGameRef<MyGame>, Tappable {
  late final SpriteButtonComponent _buttonPause;
...
...
...

_buttonPause = ButtonComponent(
        onPressed: () {
          log('_buttonPause clicked!');
          gameRef.pause();
        },
        button: SpriteComponent()
          ..sprite = await gameRef.loadSprite(
            'pause.png',
            srcPosition: Vector2(0, 0),
            srcSize: Vector2(48, 48),
          )
          ..debugMode = true
          ..decorator.replaceLast(
              PaintDecorator.tint(const Color.fromARGB(255, 255, 255, 255))));
    add(_buttonPause
      ..anchor = Anchor.topLeft
      ..size = Vector2(32, 32)
      ..position = Vector2(20 + 30, size.y / 2 + 30));
```

**Comment by spydon** [comment, marked as answer]
You shouldn't have `Tappable` on the header container and make sure that the size of the header container covers all the buttons.

**Reply by doowonee** [reply]
Thank for feedback. I add `Tappable` for debug to log event from `HeaderContainer` It doesn't work even I remove it. I'm sure `HeaderContainer` covers all the buttons area.

**Reply by spydon** [reply]
Which camera are you using, the built-in one or the camera component?

**Reply by doowonee** [reply]
Um.. I'm not sure. I just used follow camera from flame example. this is all of code about the camera.

```
// in FlamgeGame.onLoad()
camera.speed = 1;
camera.followComponent(player, worldBounds: Map.bounds);
```

And here is a screenshot

![image](https://user-images.githubusercontent.com/10105141/188277948-79498d7c-8a29-4bf1-bddc-359edcd66e93.png)

**Reply by spydon** [reply]
Hmm, just as a test, do the buttons work if you don't set the position type to viewport?

**Reply by doowonee** [reply]
Wow that's amazing it works! But I would like to make something like HUD on the screen. In my case I use following camera So any recommendation to implement? If I change the position type to game then I have to calculate position every frame and that is strange..

**Reply by spydon** [reply]
If you set the viewport position on the container instead it should work like intended I think.

**Reply by doowonee** [reply]
Sorry I don't understand you mean don't change header's positiontype and change the button's position type which child of header?

**Reply by doowonee** [reply]
I change `FlamgeGame.onLoad()` part like below

Before

```
  add(HeaderContainer()
      ..positionType = PositionType.viewport
      ..size = Vector2(size.x, 60));
```

After

```
    add(HudMarginComponent(
        margin: const EdgeInsets.only(
      top: 20,
    ))
      ..size = Vector2(size.x, 60)
      ..add(HeaderContainer()));
```

But still doesn't work. Button clicked not triggered..

**Reply by doowonee** [reply]
And I add button to flamegame without `HeaderContainer` and It works event position type is viewport. I think there is something about child component which has button with position type is viewport 

`FlamgeGame.onLoad()`

```
var _buttonPause = SpriteButtonComponent(
        onPressed: () {
          log('_buttonPause clicked 클릭');
          pause();
        },
        button: await loadSprite(
          'pause.png',
          srcPosition: Vector2(0, 0),
          srcSize: Vector2(48, 48),
        ))
      ..decorator.replaceLast(
          PaintDecorator.tint(const Color.fromARGB(255, 255, 255, 255)));
    add(_buttonPause
      ..positionType = PositionType.viewport
      ..anchor = Anchor.topLeft
      ..size = Vector2(32, 32)
      ..position = Vector2(40, 70));
```

**Reply by doowonee** [reply]
I found something! If I remove below code from `FlamgeGame.onLoad()` then buttons work. But I have to use camera follow function...

```
// camera.speed = 1;
// camera.followComponent(player, worldBounds: Map.bounds);
```

**Reply by doowonee** [reply]
I found the reason! I throughout flame game engine didn't propagate click event to child component `HeaderContainer` but it wasn't. The button was not even clicked! see screenshot.

![image](https://user-images.githubusercontent.com/10105141/188296130-7b68f1d6-d1ca-4921-8740-0ec4022da2fb.png)

`HeaderContainer` is on top of screen. I enalbed debug mode and the position of button is showing absolute position from the screen but actual position is not (maybe -200, -800). So I understand what is going on but I don't have idea to handle this.

**Reply by spydon** [reply]
Try setting the `HeaderContainer` to `positionType = PositionType.viewport` and the buttons as `positionType = PositionType.game`.

What you also could do is using the `CameraComponent` (the next generation camera) instead of the camera built into Flame game.

**Reply by doowonee** [reply]
Thank for your answer. Actually I did I change the code to use CameraComponent referncing racing game of flame example. 

But now I'm struggling about word bounds. I use `setBounds` same as world bound size but the behavior is different with prevous version of camera `camera.followComponent(player, woldbound)`. Could you give me some advice?

How can I make camera following on center of screen and when player approaches edge of screen then focus the player on edge not on center.
